client/src/stellar_error.rs

```diff
@@ -0,0 +1,244 @@
+use std::error::Error;
+use serde::{de, Deserialize, Deserializer};
+use std::fmt;
+
+/// A resource for the stellar horizon API specific error codes.
+/// These errors adhere to the Problem Details Standard
+/// https://tools.ietf.org/html/draft-ietf-appsawg-http-problem-00
+/// and a list of possible erros can be found here
+/// https://www.stellar.org/developers/horizon/reference/errors.html
+#[derive(Debug)]
+pub struct StellarError {
+    // type is a protected word
+    error_type: ErrorType,
+    title: String,
+    status: u16,
+    detail: String,
+    instance: String,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ErrorType {
+    BadRequest,
+    BeforeHistory,
+    Forbidden,
+    NotAcceptable,
+    NotFound,
+    NotImplemented,
+    RateLimitExceeded,
+    InternalServerError,
+    StaleHistory,
+    TransactionFailed,
+    TransactionMalformed,
+    OtherError(String),
+}
+
+#[derive(Deserialize)]
+struct IntermediateStellarError {
+    // type is a protected word
+    #[serde(rename = "type")] error_type: String,
```

Comment:
If the only thing being serialized here is the type, then you probably could have just implemented serde for the enum and not needed an intermediate type.

```rust
impl<'de> Deserialize<'de> for ErrorType {
  fn deserialize<'de>(d: Deserializer) -> Result {
    match String::deserialize(d)? {
      "type" => EnumType
    }
}
```

Reply:
This was a great suggestion thank you.